runtime/pkg/rduckdb/db.go

```diff
@@ -0,0 +1,1074 @@
+	}
+	err := d.writeSem.Acquire(ctx, 1)
+	if err != nil {
+		return err
+	}
+	defer d.writeSem.Release(1)
+
+	// pull latest changes from remote
+	err = d.pullFromRemote(ctx)
+	if err != nil {
+		return fmt.Errorf("rename: unable to pull from remote: %w", err)
+	}
+
+	oldMeta, err := d.catalog.tableMeta(ctx, oldName)
+	if err != nil {
+		if errors.Is(err, errNotFound) {
+			return fmt.Errorf("rename: Table %q not found", oldName)
+		}
+		return fmt.Errorf("rename: unable to get table meta: %w", err)
+	}
+
+	// copy the old table to new table
+	newVersion := newVersion()
+	err = copyDir(filepath.Join(d.localPath, newName, newVersion), filepath.Join(d.localPath, oldName, oldMeta.Version))
+	if err != nil {
+		return fmt.Errorf("rename: copy table failed: %w", err)
+	}
+
+	// rename the underlying table
+	err = renameTable(ctx, filepath.Join(d.localPath, newName, newVersion, "data.db"), oldName, newName)
+	if err != nil {
+		return fmt.Errorf("rename: rename table failed: %w", err)
+	}
+
+	d.localDirty = true
+	// sync the new table and new version
+	meta := &tableMeta{
+		Name:           newName,
+		Version:        newVersion,
+		CreatedVersion: newVersion,
+		Type:           oldMeta.Type,
+		SQL:            oldMeta.SQL,
+	}
+	if err := d.pushToRemote(ctx, newName, oldMeta, meta); err != nil {
+		return fmt.Errorf("rename: unable to replicate new table: %w", err)
+	}
+
+	// TODO :: fix this
+	// at this point db is inconsistent
+	// has both old table and new table
+
+	// drop the old table in remote
+	err = d.deleteRemote(ctx, oldName, "")
+	if err != nil {
+		return fmt.Errorf("rename: unable to delete old table %q from remote: %w", oldName, err)
+	}
+
+	// no errors after this point since background goroutine will eventually sync the local db
+
+	// update local meta for new table
+	err = d.writeTableMeta(newName, meta)
+	if err != nil {
+		d.logger.Debug("rename: error in writing table meta", slog.String("name", newName), slog.String("error", err.Error()))
+		return nil
+	}
+
+	// remove old table from local db
+	err = d.catalog.removeTable(ctx, oldName)
+	if err != nil {
+		d.logger.Debug("rename: error in removing table", slog.String("name", oldName), slog.String("error", err.Error()))
+		return nil
+	}
+	err = d.catalog.addTableVersion(ctx, newName, meta)
+	if err != nil {
+		d.logger.Debug("rename: error in adding version", slog.String("table", newName), slog.String("version", newVersion), slog.String("error", err.Error()))
+		return nil
+	}
+	d.localDirty = false
+	return nil
+}
+
+func (d *db) localDBMonitor() {
+	for {
+		select {
+		case <-d.ctx.Done():
+			return
+		case <-d.ticker.C:
+			err := d.writeSem.Acquire(d.ctx, 1)
+			if err != nil && !errors.Is(err, context.Canceled) {
+				d.logger.Error("localDBMonitor: error in acquiring write sem", slog.String("error", err.Error()))
+				continue
+			}
+			if !d.localDirty {
+				// all good
+				continue
+			}
+			err = d.pullFromRemote(d.ctx)
```

Comment:
Yes I planned to implement a ad-hoc pull when adding support for separate read and writes.

Given we have removed `ctx` calls from catalog functions, the chances of `localDirty` being true are low now. Can this be taken up separately as a follow up ? It should be a small change, need to set a channel that `localDBMonitor` listens but want to avoid making more changes in the current PR.

Reply:
Yeah if the chance is low now, then it's fine to pick it up later :)